pkg/services/settings/settingsprovider/settingsprovider.go

```diff
@@ -0,0 +1,162 @@
+package settingsprovider
+
+import (
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/registry"
+	"github.com/grafana/grafana/pkg/services/settings"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+var (
+	logger = log.New("settingsprovider")
+)
+
+func init() {
+	registry.Register(&registry.Descriptor{
+		Name:         "SettingsProvider",
+		Instance:     &Implementation{},
+		InitPriority: registry.High - 5,
+	})
+}
+
+type settingsBag map[string]map[string]string
+
+type Implementation struct {
+	sync.RWMutex
+	FileCfg  *setting.Cfg       `inject:""`
+	SQLStore *sqlstore.SQLStore `inject:""`
```

Comment:
I decided to build this new abstraction on top of these two elements because they have a lot of logics (all the overrides from CLI args, env vars, how to build the db connection string, etc) that will be complex to migrate to this new piece, but I guess at the end those will be partly moved into this new component. Does that make sense? 🤔